src/storage/src/mem/mem_bucket.rs

```
@@ -0,0 +1,106 @@
+// Copyright 2021 The Engula Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{collections::HashMap, sync::Arc};
+
+use futures::stream;
+use tokio::sync::Mutex;
+
+use super::mem_object::MemObject;
+use crate::{
+    async_trait, Error, Result, StorageBucket, StorageObject, StorageObjectUploader, Stream,
+};
+
+type Objects = Arc<Mutex<HashMap<String, MemObject>>>;
+
+#[derive(Clone)]
+pub struct MemBucket {
+    objects: Objects,
+}
+
+impl MemBucket {
+    pub fn new() -> MemBucket {
+        MemBucket {
+            objects: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+#[async_trait]
+impl StorageBucket for MemBucket {
+    async fn object(&self, name: &str) -> Result<Box<dyn StorageObject>> {
+        let objects = self.objects.lock().await;
+        match objects.get(name) {
+            Some(object) => Ok(Box::new(object.clone())),
+            None => Err(Error::NotFound(format!("object '{}'", name))),
+        }
+    }
+
+    async fn list_objects(&self) -> Stream<Result<String>> {
+        let objects = self.objects.lock().await;
+        let object_names = objects
+            .keys()
+            .cloned()
+            .map(Ok)
+            .collect::<Vec<Result<String>>>();
+        Box::new(stream::iter(object_names))
+    }
+
+    async fn upload_object(&self, name: &str) -> Box<dyn StorageObjectUploader> {
```

maybe we need to detect the state of the bucket that owned this object?

e.g. two people A and B
---- A: get bucket "bucket-test" that exists
---- B: empty and delete "bucket-test"
---- A:  try to upload object into "bucket-test"  <=== should return an error, because the object can not download?

---

I think it depends. If we mimic file system behaviors on Linux, it is OK to use an opened but deleted directory/file. Once the directory is closed, all data is gone. But I think we should follow the practice of different object storages here.